# Design an application for public health

### Context:
- The agency "Santé publique France" has launched a call for projects to find innovative ideas for applications related to food. You want to participate and propose an idea for an application.

#### Mission

1) Process the dataset, by:
    Thinking about an application idea.
    Identifying variables that are relevant to future processing and necessary for your application idea.
    Cleaning the data by :
    Highlighting any missing values, with at least 3 processing methods appropriate for the variables involved, identifying and quantifying the possible outliers of each variable.
    Automating these processes to avoid repeating these operations.
    
    
2) Throughout the analysis, produce visualizations to better understand the data. Perform a univariate analysis for each variable of interest to summarize its behavior.


3) Confirm or refute hypotheses using multivariate analysis. Perform appropriate statistical tests to verify the significance of the results.


4) Justify your application idea. Identify arguments to justify the feasibility (or not) of the application based on the Open Food Facts data.

## Data source and details
The CSV data file for this project is a public datasets from Open Food Facts - World (https://world.openfoodfacts.org/) and made available by the OpenClassroms. 

## PART 1 - Data Cleaning

## 1.0 Python libraries and datasets

### 1.1 Importation of python libraries 

In [1]:
# importation of libraries 
import pandas as pd # for data loading and processing
import numpy as np # for linear algebra

### 1.2 Importation of dataset

In [2]:
# loading of dataset to the Pandas DataFrame

Data = pd.read_csv('openfoodfacts.csv', sep = '\t', engine='python', on_bad_lines = 'skip')
Data.head() #display data

code                                                url  \
0  0000000003087  http://world-fr.openfoodfacts.org/produit/0000...   
1  0000000004530  http://world-fr.openfoodfacts.org/produit/0000...   
2  0000000004559  http://world-fr.openfoodfacts.org/produit/0000...   
3  0000000016087  http://world-fr.openfoodfacts.org/produit/0000...   
4  0000000016094  http://world-fr.openfoodfacts.org/produit/0000...   

                      creator   created_t      created_datetime  \
0  openfoodfacts-contributors  1474103866  2016-09-17T09:17:46Z   
1             usda-ndb-import  1489069957  2017-03-09T14:32:37Z   
2             usda-ndb-import  1489069957  2017-03-09T14:32:37Z   
3             usda-ndb-import  1489055731  2017-03-09T10:35:31Z   
4             usda-ndb-import  1489055653  2017-03-09T10:34:13Z   

  last_modified_t last_modified_datetime                    product_name  \
0      1474103893   2016-09-17T09:18:13Z              Farine de blé noir   
1      1489069957   2017-03-09T14:32:37Z  Banana Chips Sweetened (Whole)   
2      1489069957   2017-03-09T14:32:37Z                         Peanuts   
3      1489055731   2017-03-09T10:35:31Z          Organic Salted Nut Mix   
4      1489055653   2017-03-09T10:34:13Z                 Organic Polenta   

  generic_name quantity  ... ph_100g fruits-vegetables-nuts_100g  \
0          NaN      1kg  ...     NaN                         NaN   
1          NaN      NaN  ...     NaN                         NaN   
2          NaN      NaN  ...     NaN                         NaN   
3          NaN      NaN  ...     NaN                         NaN   
4          NaN      NaN  ...     NaN                         NaN   

  collagen-meat-protein-ratio_100g cocoa_100g chlorophyl_100g  \
0                              NaN        NaN             NaN   
1                              NaN        NaN             NaN   
2                              NaN        NaN             NaN   
3                              NaN        NaN             NaN   
4                              NaN        NaN             NaN   

  carbon-footprint_100g nutrition-score-fr_100g nutrition-score-uk_100g  \
0                   NaN                     NaN                     NaN   
1                   NaN                    14.0                    14.0   
2                   NaN                     0.0                     0.0   
3                   NaN                    12.0                    12.0   
4                   NaN                     NaN                     NaN   

  glycemic-index_100g water-hardness_100g  
0                 NaN                 NaN  
1                 NaN                 NaN  
2                 NaN                 NaN  
3                 NaN                 NaN  
4                 NaN                 NaN  

[5 rows x 162 columns]

### 1.3 Pre-exploratory analysis

In [3]:
# Summary of dataset

Data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 320767 entries, 0 to 320766
Columns: 162 entries, code to water-hardness_100g
dtypes: float64(106), object(56)
memory usage: 396.5+ MB


In [4]:
# Dimension of the dataset
Data.shape

(320767, 162)

 - There are total 320767 food products with 162 features listed in the open food facts dataset

In [5]:
#data type info
Data.dtypes

code                        object
url                         object
creator                     object
created_t                   object
created_datetime            object
                            ...   
carbon-footprint_100g      float64
nutrition-score-fr_100g    float64
nutrition-score-uk_100g    float64
glycemic-index_100g        float64
water-hardness_100g        float64
Length: 162, dtype: object

 - The dataset consists of both categorical and numerical data types

In [6]:
# check for duplicated rows
#Data[Data.duplicated('code')]
Data.duplicated().sum()

0

 - There is no duplicated product in the dataset

In [7]:
# check for any missing data in the df (display in descending order)
Data.isna().sum().sort_values(ascending=False)

water-hardness_100g                      320767
no_nutriments                            320767
ingredients_that_may_be_from_palm_oil    320767
nutrition_grade_uk                       320767
nervonic-acid_100g                       320767
                                          ...  
created_datetime                              9
created_t                                     3
creator                                       2
last_modified_datetime                        0
last_modified_t                               0
Length: 162, dtype: int64

 - The dataset consist of thousands of missing data and some of the feaures/variables are empty 

### 2.0 Data cleaning and preparation

#### 2.1. Drop variables with a high percent of missing values 
- keeping variable with high percentage of missing will disturb the final data interpretations

In [8]:
# Remove columns with more than 80% NaN threshold

Data = Data[Data.columns[Data.isna().sum()/len(Data) <0.8]]

In [9]:
# Display names of the remaining columns
Data.columns 

Index(['code', 'url', 'creator', 'created_t', 'created_datetime',
       'last_modified_t', 'last_modified_datetime', 'product_name', 'quantity',
       'packaging', 'packaging_tags', 'brands', 'brands_tags', 'categories',
       'categories_tags', 'categories_fr', 'countries', 'countries_tags',
       'countries_fr', 'ingredients_text', 'serving_size', 'additives_n',
       'additives', 'additives_tags', 'additives_fr',
       'ingredients_from_palm_oil_n',
       'ingredients_that_may_be_from_palm_oil_n', 'nutrition_grade_fr',
       'pnns_groups_1', 'pnns_groups_2', 'states', 'states_tags', 'states_fr',
       'main_category', 'main_category_fr', 'image_url', 'image_small_url',
       'energy_100g', 'fat_100g', 'saturated-fat_100g', 'trans-fat_100g',
       'cholesterol_100g', 'carbohydrates_100g', 'sugars_100g', 'fiber_100g',
       'proteins_100g', 'salt_100g', 'sodium_100g', 'vitamin-a_100g',
       'vitamin-c_100g', 'calcium_100g', 'iron_100g',
       'nutrition-score-fr_100g', 

In [10]:
# Filter relevant numerical data columns

# numerical columns
cols_1 = ['energy_100g', 'fat_100g', 'saturated-fat_100g', 'carbohydrates_100g', 'sugars_100g', 'fiber_100g',
       'proteins_100g', 'salt_100g', 'sodium_100g', 'nutrition-score-fr_100g', 'nutrition-score-uk_100g']

Cols_imp_1 = Data[cols_1]

# Select relevant categorical columns

Cols_2 = ['code', 'product_name', 'pnns_groups_1', 'pnns_groups_2', 'nutrition_grade_fr'] 
Cols_imp_2 = Data[Cols_2]

# Concatenate Cols_imp_1 and Cols_imp_2

Data_N = pd.concat([Cols_imp_2, Cols_imp_1], axis = 1) 

#### 2.2. Remove rows with missing categorical values 
- since categorical variables cannot be replaced by statistical imputation methods, it's appropriate to drop them

In [11]:
# create list of columns of caterogical
#Col_nm = ['code', 'product_name', 'pnns_groups_1', 'pnns_groups_2', 'nutrition_grade_fr'] 


# Drop rows of categorical variables/columns having the missing values 
Data_df = Data_N.dropna(subset = Cols_2)

# Drop pnns_groups_1 and pnns_group_2 with unknowns

Data_df = Data_df.drop(Data_df.loc[Data_df['pnns_groups_1'] == 'unknown', 'pnns_groups_1'].index)

Data_df = Data_df.drop(Data_df.loc[Data_df['pnns_groups_2'] == 'unknown', 'pnns_groups_2'].index)

#### 2.3. replace outliers

In [12]:
# source https://ib.bioninja.com.au/options/option-d-human-physiology/d1-human-nutrition/energy-sources.html) 
# (source - https://www.myfooddata.com/articles/foods-high-in-dietary-fiber.php)

# energy
# Replace values > 4000 KJ (maximum in a 100g product) in energy_100g with NaN

energy = Data_df.loc[Data_df['energy_100g'] > 4000, 'energy_100g'].tolist()
Data_df.replace(energy, np.nan, inplace = True)

# saturated fat > fat
fat = Data_df.loc[Data_df['saturated-fat_100g'] > Data_df['fat_100g'],'saturated-fat_100g'].tolist()
Data_df.replace(fat, np.nan, inplace = True)

# Sugar > carbohydrate
sugars = Data_df.loc[Data_df['sugars_100g'] > Data_df['carbohydrates_100g'],'sugars_100g'].tolist()
Data_df.replace(sugars, np.nan, inplace = True)

# Proteins > 60  Data_df[Data_df['proteins_100g'] > 60] 
protein = Data_df.loc[Data_df['proteins_100g'] > 60,'proteins_100g'].tolist()
Data_df.replace(protein, np.nan, inplace = True)

# fiber > 20 : Data-df[Data_df['fiber_100g'] > 20]
fiber = Data_df.loc[Data_df['fiber_100g'] > 20,'fiber_100g'].tolist()
Data_df.replace(fiber, np.nan, inplace = True)

# fat > 100 : Data[Data['fat_100g'] > 100]
fat = Data_df.loc[Data_df['fat_100g'] > 100,'fat_100g'].tolist()
Data_df.replace(fat, np.nan, inplace = True)
                  
# Salt > 30: Data_df[Data_df['salt_100g'] > 30] 
salt = Data_df.loc[Data_df['salt_100g'] > 30,'salt_100g'].tolist()
Data_df.replace(salt, np.nan, inplace = True)

#### 2.4. fill the remaining missing numerical (quantitative) data with the mean of each group

In [13]:
Data_df.isna().sum()/len(Data_df)

code                       0.000000
product_name               0.000000
pnns_groups_1              0.000000
pnns_groups_2              0.000000
nutrition_grade_fr         0.000000
energy_100g                0.020730
fat_100g                   0.504537
saturated-fat_100g         0.442507
carbohydrates_100g         0.407402
sugars_100g                0.447868
fiber_100g                 0.609833
proteins_100g              0.386196
salt_100g                  0.218136
sodium_100g                0.049937
nutrition-score-fr_100g    0.520144
nutrition-score-uk_100g    0.522030
dtype: float64

In [14]:
# Extract the list of numerical columns

Data_df_nc = Data_df.drop(Cols_2, axis = 1)

# create list of numerical columns with missing data

col_nc = Data_df_nc.columns[0:].tolist()

# group by pnns_group and fill with Nan with the group mean
Data_nc = Data_df[col_nc].fillna(Data_df.groupby('pnns_groups_1')[col_nc].transform('mean'))

# concat dataframes of cleaned numerical and categorical dataframes
Data_clean = pd.concat([Data_df[Cols_2], Data_nc], axis = 1)
Data_clean

code                         product_name  \
174     0000000290616                         Salade Cesar   
180     0000001938067         Chaussons tressés aux pommes   
185     0000008237798                      Quiche Lorraine   
218     0000020424640                   Belgische Pralinen   
223     0000027533048             Luxury Christmas Pudding   
...               ...                                  ...   
320688       96092521    Santa Cruz Chilli & Lime Dressing   
320697       96118580       Fisherman's Friend Miel-Citron   
320729  9778564312457  Nature cheddar and red onion crisps   
320746  9847548283004  Tartines craquantes bio au sarrasin   
320758       99111250                   Thé vert Earl grey   

                pnns_groups_1          pnns_groups_2 nutrition_grade_fr  \
174     Fruits and vegetables             Vegetables                  c   
180             Sugary snacks     Biscuits and cakes                  c   
185           Composite foods  Pizza pies and quiche                  b   
218             Sugary snacks                 Sweets                  e   
223             Sugary snacks     Biscuits and cakes                  c   
...                       ...                    ...                ...   
320688         Fat and sauces   Dressings and sauces                  c   
320697          Sugary snacks                 Sweets                  b   
320729           Salty snacks             Appetizers                  c   
320746   Cereals and potatoes                  Bread                  a   
320758              Beverages  Non-sugared beverages                  c   

        energy_100g   fat_100g  saturated-fat_100g  carbohydrates_100g  \
174     1210.000000   3.968002            7.000000           17.319731   
180     1090.000000  10.700000            9.124302           38.700000   
185      478.000000   6.790000            2.860000            7.860000   
218     2257.000000  33.300000           21.100000           60.565619   
223     1284.000000   7.000000            3.600000           54.200000   
...             ...        ...                 ...                 ...   
320688   660.000000  46.570049           14.101074            7.776859   
320697  1031.000000   1.300000            1.280000           95.310000   
320729  2155.000000  33.420577            3.400000           47.357783   
320746  1643.000000   2.800000            2.403335           74.800000   
320758   287.684571   0.648130            0.367114           13.386056   

        sugars_100g  fiber_100g  proteins_100g  salt_100g  sodium_100g  \
174        0.000000    2.950240       2.231402   0.511223     0.850000   
180       24.700000    3.520498       3.330000   0.647000     0.255000   
185        0.714000    0.357000       5.360000   0.499000     0.196000   
218       51.500000    3.520498       4.700000   0.090000     0.035433   
223       45.109926    4.700000       5.877508   0.199898     0.078700   
...             ...         ...            ...        ...          ...   
320688    16.700000    0.831137       1.941916   0.541020     0.213000   
320697    45.109926    1.470000       0.004000   0.364057     0.000394   
320729     1.800000    5.190905       7.800000   1.350000     0.531496   
320746     2.600000    5.900000      13.000000   0.680000     0.267717   
320758    11.876192    0.321452       0.810621   0.025400     0.028669   

        nutrition-score-fr_100g  nutrition-score-uk_100g  
174                    6.000000                 6.000000  
180                    9.000000                 9.000000  
185                    2.131309                 2.136796  
218                   11.977283                11.978147  
223                   10.000000                10.000000  
...                         ...                      ...  
320688                 6.000000                 6.000000  
320697                11.977283                11.978147  
320729                 9.000000                 9.000000  
320

### 3.0 Quality check of the cleaned data

In [15]:
# Statistical verification

Data_clean.describe()

energy_100g      fat_100g  saturated-fat_100g  carbohydrates_100g  \
count  50363.000000  50363.000000        50363.000000        50363.000000   
mean    1106.204426     13.195526            5.620969           28.251899   
std      789.105658     15.980304            7.052180           26.304260   
min        0.000000      0.000000            0.000000            0.000000   
25%      394.000000      3.400000            1.100000            7.776859   
50%     1006.000000      7.719382            3.100000           13.762320   
75%     1663.000000     19.034108            8.599619           57.316974   
max     4000.000000    100.000000          210.000000          190.000000   

        sugars_100g    fiber_100g  proteins_100g     salt_100g   sodium_100g  \
count  50363.000000  50363.000000   50363.000000  50363.000000  50363.000000   
mean      13.940586      2.301679       6.748702      0.831923      0.337414   
std       18.320195      2.487244       5.260810      1.130462      0.973476   
min       -0.100000      0.000000       0.000000      0.000000      0.000000   
25%        2.649772      0.294567       2.231402      0.090000      0.028669   
50%        6.111708      1.690000       6.500000      0.635000      0.200787   
75%       13.736174      3.520498       9.463442      1.220000      0.472441   
max      134.000000     19.900000      59.100000     29.470000    117.165354   

       nutrition-score-fr_100g  nutrition-score-uk_100g  
count             50363.000000             50363.000000  
mean                  5.930124                 4.915013  
std                   6.446020                 6.708562  
min                 -15.000000               -15.000000  
25%                   0.000000                -1.000000  
50%                   7.048086                 6.299840  
75%                  11.000000                11.000000  
max                  18.000000                18.000000

- Remove negative values (except nutrition scores) and other outliers after imputation

In [16]:
# saturated fat > fat
fat_2 = Data_clean.loc[Data_clean['saturated-fat_100g'] > Data_clean['fat_100g'],'saturated-fat_100g'].index
Data_clean.drop(fat_2, inplace = True)

# sugar with negative values
sugars_1 = Data_clean.loc[Data_clean['sugars_100g'] < 0, 'sugars_100g'].index
Data_clean.drop(sugars_1, inplace = True)

# Sugar > carbohydrate
sugars_2 = Data_clean.loc[Data_clean['sugars_100g'] > Data_clean['carbohydrates_100g'],'sugars_100g'].index
Data_clean.drop(sugars_2, inplace = True)

In [17]:
Data_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 40421 entries, 180 to 320758
Data columns (total 16 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   code                     40421 non-null  object 
 1   product_name             40421 non-null  object 
 2   pnns_groups_1            40421 non-null  object 
 3   pnns_groups_2            40421 non-null  object 
 4   nutrition_grade_fr       40421 non-null  object 
 5   energy_100g              40421 non-null  float64
 6   fat_100g                 40421 non-null  float64
 7   saturated-fat_100g       40421 non-null  float64
 8   carbohydrates_100g       40421 non-null  float64
 9   sugars_100g              40421 non-null  float64
 10  fiber_100g               40421 non-null  float64
 11  proteins_100g            40421 non-null  float64
 12  salt_100g                40421 non-null  float64
 13  sodium_100g              40421 non-null  float64
 14  nutrition-score-fr_

In [18]:
Data_clean.shape

(40421, 16)

 - The dimension of the dataset is reduced from 32076 x 162 to 40421 x 16 after cleaning

In [19]:
# Missing values in the final dataframe
Data_clean.isna().sum()

code                       0
product_name               0
pnns_groups_1              0
pnns_groups_2              0
nutrition_grade_fr         0
energy_100g                0
fat_100g                   0
saturated-fat_100g         0
carbohydrates_100g         0
sugars_100g                0
fiber_100g                 0
proteins_100g              0
salt_100g                  0
sodium_100g                0
nutrition-score-fr_100g    0
nutrition-score-uk_100g    0
dtype: int64

 - All missing values corrected removed and replaced

In [20]:
# Final statistical check and value range
Data_clean.describe()

energy_100g      fat_100g  saturated-fat_100g  carbohydrates_100g  \
count  40421.000000  40421.000000        40421.000000        40421.000000   
mean    1142.177714     14.529826            4.919651           30.708133   
std      812.873006     16.819655            6.582719           26.666932   
min        0.000000      0.000000            0.000000            0.000000   
25%      410.000000      3.968002            0.400000            7.800000   
50%     1016.000000      8.600000            3.065319           15.500000   
75%     1790.000000     19.034108            8.000000           57.316974   
max     4000.000000    100.000000          100.000000          139.000000   

        sugars_100g    fiber_100g  proteins_100g     salt_100g   sodium_100g  \
count  40421.000000  40421.000000   40421.000000  40421.000000  40421.000000   
mean      14.502450      2.480131       6.771932      0.823504      0.336247   
std       18.952494      2.481954       5.354204      1.135229      1.059603   
min        0.000000      0.000000       0.000000      0.000000      0.000000   
25%        2.600000      0.321452       2.231402      0.070000      0.019685   
50%        6.111708      2.157020       6.400000      0.584200      0.207000   
75%       13.736174      3.520498       9.463442      1.200000      0.459000   
max      134.000000     19.900000      59.100000     29.470000    117.165354   

       nutrition-score-fr_100g  nutrition-score-uk_100g  
count             40421.000000             40421.000000  
mean                  6.047936                 4.959839  
std                   6.452560                 6.792263  
min                 -15.000000               -15.000000  
25%                   0.000000                -1.095729  
50%                   7.092639                 6.591861  
75%                  11.977283                11.000000  
max                  18.000000                18.000000

In [21]:
#Export and save the cleaned final dataframe to csv file for exploratory analysis

Data_clean = Data_clean.set_index('code')
Data_clean.to_csv("Clean_Data_P3.csv") 